lib/FrontendTool/FrontendTool.cpp

```diff
@@ -314,47 +311,43 @@ static bool writeSIL(SILModule &SM, const PrimarySpecificPaths &PSPs,
                   PSPs.OutputFilename, opts.EmitSortedSIL);
 }
 
-static bool printAsObjCIfNeeded(StringRef outputPath, ModuleDecl *M,
-                                StringRef bridgingHeader, bool moduleIsPublic) {
-  using namespace llvm::sys;
-
+/// A wrapper around swift::atomicallyWritingToFile that handles diagnosing any
+/// filesystem errors and ignores empty output paths.
+static bool atomicallyWritingToTextFile(
+    StringRef outputPath, DiagnosticEngine &diags,
+    llvm::function_ref<bool(llvm::raw_pwrite_stream &)> action) {
   if (outputPath.empty())
     return false;
 
-  clang::CompilerInstance Clang;
-
-  std::string tmpFilePath;
-  std::error_code EC;
-  std::unique_ptr<llvm::raw_pwrite_stream> out =
-    Clang.createOutputFile(outputPath, EC,
-                           /*Binary=*/false,
-                           /*RemoveFileOnSignal=*/true,
-                           /*BaseInput=*/"",
-                           path::extension(outputPath),
-                           /*UseTemporary=*/true,
-                           /*CreateMissingDirectories=*/false,
-                           /*ResultPathName=*/nullptr,
-                           &tmpFilePath);
-
-  if (!out) {
-    M->getASTContext().Diags.diagnose(SourceLoc(), diag::error_opening_output,
-                                      tmpFilePath, EC.message());
-    return true;
-  }
-
-  auto requiredAccess = moduleIsPublic ? AccessLevel::Public
-                                       : AccessLevel::Internal;
-  bool hadError = printAsObjC(*out, M, bridgingHeader, requiredAccess);
-  out->flush();
-
-  EC = swift::moveFileIfDifferent(tmpFilePath, outputPath);
+  bool actionFailed = false;
+  std::error_code EC =
+      swift::atomicallyWritingToFile(outputPath, /*binary*/false,
+                                     [&](llvm::raw_pwrite_stream &out) {
+    actionFailed = action(out);
+  });
   if (EC) {
-    M->getASTContext().Diags.diagnose(SourceLoc(), diag::error_opening_output,
-                                      outputPath, EC.message());
+    diags.diagnose(SourceLoc(), diag::error_opening_output,
+                   outputPath, EC.message());
     return true;
   }
+  return actionFailed;
+}
+
+static bool printAsObjCIfNeeded(StringRef outputPath, ModuleDecl *M,
```

Comment:
The name says "IfNeeded" but the function has no "if" test in it. I presume that it has to do with an empty output path? Some sort of reformulation could be clearer.

Reply:
Same for `printModuleInterfaceIfNeeded` below.

Reply:
My preference would be for an empty output path to be an error, and to only call the function with a non-empty output path. But one could instead append "IgnoringEmptyOutputPath" to the function name.

Reply:
That's definitely too verbose—it makes the call site too hard to read.

I can push the empty string checks back out of these functions, but it was your refactoring of performCompile that put them inside to begin with. :-/

Reply:
I think that the function name is best when consistent with the semantics. If you don't want the verbose name for the atomic write function, how about only calling it with a non-empty output path?

Reply:
Would you prefer the checks to be:
- ~in the helper function~
- in `printAsObjC[IfNeeded]` and `emitModuleInterface[IfNeeded]`
- at the call sites for `printAsObjC` and `emitModuleInterface`?

Reply:
Thanks for asking. It's a question of centralizing the invariants. If there is only one call to printAsObjC, then I would put the test at the call site. If there are two, I would put the test in printAsObjCIfNeeded and have just test and call printAsObjC (or whatever). Same for emitModuleInterface[IfNeeded]. Does that make sense?

Reply:
There's only one call to all of these (they were all inline in `performCompile` at one point), so I'll put it outside, then.

Reply:
Great!

Reply:
Oops, I take it back. There are two calls. Hm. So you're saying it's better to have the test in `printAsObjCIfNeeded` rather than sink it down to the helper function, since we haven't thought of a way to encode it in the helper function's name without me thinking it's too verbose?

Reply:
Yes, that’s right. One place, so someday we can’t change one and miss the other. Tx